File: src/test/kotlin/model/StatesTest.kt
package model

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

internal class StatesTest {
    private lateinit var states: States

    @BeforeEach
    fun setUp() {
        states = States()
    }

    @Test
    fun `test makeState creates a new forest with two nodes`() {
        states.makeState("someCity", "someState")

        assertEquals(2, states.states.predecessor.size)
        assertEquals("someState", states.states.predecessor["someCity"])
        assertEquals("someState", states.states.predecessor["someState"])

        assertEquals(2, states.states.rank.size)
        assertEquals(0, states.states.rank["someCity"])
        assertEquals(1, states.states.rank["someState"])
    }

    @Test
    fun `test union merges two states and updates root node`() {
        states.makeState("cityA", "stateA")
        states.makeState("cityB", "stateB")

        states.union("stateB", "stateA")

        assertEquals("stateA", states.findState("cityA"))
        assertEquals("stateA", states.findState("cityB"))
    }

    @Test
    fun `test union merges two states and updates ranks`() {
        states.makeState("cityA", "stateA")
        states.makeState("cityB", "stateB")

        states.union("stateB", "stateA")

        assertEquals(0, states.states.rank["cityA"])
        assertEquals(0, states.states.rank["cityB"])
        assertEquals(2, states.states.rank["stateA"])
        assertEquals(1, states.states.rank["stateB"])
    }

    @Test
    fun `test union by rank`() {
        states.makeState("cityA", "stateA")
        states.makeState("cityB", "stateB")
        states.makeState("cityC", "stateC")

        states.union("stateB", "stateA")

        /**
         * Instead of using stateC as the new root node,
         * it chooses stateA as the new root node,
         * because higher ranks will always win over dominantState.
         */
        states.union("stateA", "stateC")

        assertEquals("stateA", states.states.predecessor["stateA"])
        assertEquals("stateA", states.states.predecessor["stateB"])
        assertEquals("stateA", states.states.predecessor["stateC"])
    }

    @Test
    fun `test union is always merging root nodes`() {
        states.makeState("someCity", "someState")
        states.makeState("anotherCity", "anotherState")

        assertEquals("someState", states.states.predecessor["someCity"])
        assertEquals("someState", states.states.predecessor["someState"])
        assertEquals("anotherState", states.states.predecessor["anotherCity"])
        assertEquals("anotherState", states.states.predecessor["anotherState"])

        states.union("someCity", "anotherCity")

        assertEquals("someState", states.states.predecessor["someCity"])
        assertEquals("anotherState", states.states.predecessor["someState"])
        assertEquals("anotherState", states.states.predecessor["anotherCity"])
        assertEquals("anotherState", states.states.predecessor["anotherState"])
    }

    @Test
    fun `test findState with path compression`() {
        states.makeState("cityA", "stateA")
        states.makeState("cityB", "stateB")
        states.makeState("cityC", "stateC")

        states.union("stateB", "stateA")
        states.union("stateC", "stateB")

        assertEquals("stateC", states.states.predecessor["cityC"])
        assertEquals("stateA", states.findState("cityC"))
        assertEquals("stateA", states.states.predecessor["cityC"])
    }

    @Test
    fun `test findState fails for unknown city`() {
        assertThrows<NullPointerException> {
            states.findState("nonExistingCity")
        }
    }
}

File: src/__tests__/model/states.test.ts
import { describe, beforeEach, expect, test } from 'vitest';
import { States } from '$lib/model/states';

describe('States', () => {
  let states: States;

  beforeEach(() => {
    states = new States();
  });

  test('test makeState creates a new forest with two nodes', () => {
    states.makeState('someCity', 'someState');

    expect(Object.keys(states.states.predecessor).length).toBe(2);
    expect(states.states.predecessor['someCity']).toBe('someState');
    expect(states.states.predecessor['someState']).toBe('someState');

    expect(Object.keys(states.states.rank).length).toBe(2);
    expect(states.states.rank['someCity']).toBe(0);
    expect(states.states.rank['someState']).toBe(1);
  });

  test('test union merges two states and updates root node', () => {
    states.makeState('cityA', 'stateA');
    states.makeState('cityB', 'stateB');

    states.union('stateB', 'stateA');

    expect(states.findState('cityA')).toBe('stateA');
    expect(states.findState('cityB')).toBe('stateA');
  });

  test('test union merges two states and updates ranks', () => {
    states.makeState('cityA', 'stateA');
    states.makeState('cityB', 'stateB');

    states.union('stateB', 'stateA');

    expect(states.states.rank['cityA']).toBe(0);
    expect(states.states.rank['cityB']).toBe(0);
    expect(states.states.rank['stateA']).toBe(2);
    expect(states.states.rank['stateB']).toBe(1);
  });

  test('test union by rank', () => {
    states.makeState('cityA', 'stateA');
    states.makeState('cityB', 'stateB');
    states.makeState('cityC', 'stateC');

    states.union('stateB', 'stateA');

    // Instead of using stateC as the new root node,
    // it chooses stateA as the new root node,
    // because higher ranks will always win over dominantState.
    states.union('stateA', 'stateC');

    expect(states.states.predecessor['stateA']).toBe('stateA');
    expect(states.states.predecessor['stateB']).toBe('stateA');
    expect(states.states.predecessor['stateC']).toBe('stateA');
  });

  test('test union is always merging root nodes', () => {
    states.makeState('someCity', 'someState');
    states.makeState('anotherCity', 'anotherState');

    expect(states.states.predecessor['someCity']).toBe('someState');
    expect(states.states.predecessor['someState']).toBe('someState');
    expect(states.states.predecessor['anotherCity']).toBe('anotherState');
    expect(states.states.predecessor['anotherState']).toBe('anotherState');

    states.union('someCity', 'anotherCity');

    expect(states.states.predecessor['someCity']).toBe('someState');
    expect(states.states.predecessor['someState']).toBe('anotherState');
    expect(states.states.predecessor['anotherCity']).toBe('anotherState');
    expect(states.states.predecessor['anotherState']).toBe('anotherState');
  });

  test('test findState with path compression', () => {
    states.makeState('cityA', 'stateA');
    states.makeState('cityB', 'stateB');
    states.makeState('cityC', 'stateC');

    states.union('stateB', 'stateA');
    states.union('stateC', 'stateB');

    expect(states.states.predecessor['cityC']).toBe('stateC');
    expect(states.findState('cityC')).toBe('stateA');
    expect(states.states.predecessor['cityC']).toBe('stateA');
  });

  test('test findState fails for unknown city', () => {
    expect(states.findState('nonExistingCity')).toBe(undefined);
  });

});
